Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm

```diff
@@ -994,36 +1001,118 @@
+    }();
+
+    auto applyLinkDecorationFiltering = [frame = m_frame](const URL& url) {
+        if (!frame || !frame->coreLocalFrame())
+            return url;
+        RefPtr document = frame->coreLocalFrame()->document();
+        if (!document)
+            return url;
+        RefPtr page = document->page();
+        if (!page)
+            return url;
+        return page->applyLinkDecorationFiltering(url, LinkDecorationFilteringTrigger::Copy);
+    };
+
+    std::optional<PasteboardWebContent> pasteboardContent;
+    std::optional<PasteboardURL> pasteboardURL;
+
+    auto ensureContent = [originIdentifier](std::optional<PasteboardWebContent>& content) -> decltype(content) {
+        if (!content)
+            content = PasteboardWebContent { .contentOrigin = originIdentifier, .canSmartCopyOrDelete = false };
+        return content;
+    };
 
     for (auto&& [data, type] : WTFMove(pasteboardItems)) {
-        // We don't expect the data for any items to be empty, but aren't completely sure.
-        // Avoid crashing in the SharedMemory constructor in release builds if we're wrong.
-        ASSERT([data length]);
-        if (![data length])
+        if (![data length]) {
+            ASSERT_NOT_REACHED();
             continue;
+        }
 
-        if ([type isEqualToString:legacyStringPasteboardType()] || [type isEqualToString:NSPasteboardTypeString]) {
-            auto plainTextString = adoptNS([[NSString alloc] initWithData:data.get() encoding:NSUTF8StringEncoding]);
-            webProcess.parentProcessConnection()->sendSync(Messages::WebPasteboardProxy::SetPasteboardStringForType(pasteboardName, type.get(), plainTextString.get(), pageIdentifier), 0);
-        } else {
-            auto buffer = SharedBuffer::create(data.get());
-            webProcess.parentProcessConnection()->sendSync(Messages::WebPasteboardProxy::SetPasteboardBufferForType(pasteboardName, type.get(), WTFMove(buffer), pageIdentifier), 0);
+        if ([type isEqualToString:htmlPasteboardType()])
+            ensureContent(pasteboardContent)->dataInHTMLFormat = String { adoptNS([[NSString alloc] initWithData:data.get() encoding:NSUTF8StringEncoding]).autorelease() };
+        else if ([type isEqualToString:rtfPasteboardType()])
+            ensureContent(pasteboardContent)->dataInRTFFormat = SharedBuffer::create(data.get());
+        else if ([type isEqualToString:stringPasteboardType()])
+            ensureContent(pasteboardContent)->dataInStringFormat = String { adoptNS([[NSString alloc] initWithData:data.get() encoding:NSUTF8StringEncoding]).autorelease() };
+        else if ([type isEqualToString:urlPasteboardType()]) {
+            URL url { [NSURL URLWithDataRepresentation:data.get() relativeToURL:nil] };
+            URL sanitizedURL { applyLinkDecorationFiltering(url) };
+            pasteboardURL = PasteboardURL {
+                .url = sanitizedURL,
+                .title = sanitizedURL.string(),
```

Comment:
I think we should just omit the title here — it's meant to be the text of the link (like `link` in the following [link](https://github.com) rather than the URL string).

Reply:
I did initially omit the title, but we end up advertising an unexpected title with the `public.url-name` pasteboard type, which regresses pasting from MiniBrowser into Notes.app, for example. (links show up as [/](/))

```
    NSString *title = (NSString *)pasteboardURL.title;
    if (![title length]) {
        title = [[cocoaURL path] lastPathComponent];  ---> title = '/'
        if (![title length])
            title = userVisibleString;
    }

    ...

    // Puts `/` on the pasteboard.
    newChangeCount = platformStrategies()->pasteboardStrategy()->setStringForType(title, WebURLNamePboardType, pasteboardName, context);
```

Reply:
Alternatively I could teach `Pasteboard::write(const PasteboardURL&)` not to unconditionally write for type `WebURLNamePboardType` (i.e. not when invoked from the plugin)

Reply:
Got it! Falling back to the absolute URL string seems fine, then. Ideally we'd be able to get the title from the annotation text itself, but (from what I could tell) there isn't a great way to achieve that :P